networkit/cpp/randomization/Curveball.h

```diff
@@ -0,0 +1,52 @@
+/*
+ * Curveball.h
+ *
+ *  Created on: 26.05.2018
+ *      Author:  Hung Tran <htran@ae.cs.uni-frankfurt.de>, Manuel Penschuck <networkit@manuel.jetzt>
+ */
+#ifndef RANDOMIZATION_CURVEBALL_H
+#define RANDOMIZATION_CURVEBALL_H
+
+#include <memory>
+#include <utility>
+
+#include "../Globals.h"
+#include "../base/Algorithm.h"
+#include "../graph/Graph.h"
+
+namespace NetworKit {
+
+// pImpl
+namespace CurveballDetails { struct CurveballIM; }
+
+class Curveball : public Algorithm {
+public:
+
+	explicit Curveball(const NetworKit::Graph &G);
+
+	virtual ~Curveball();
+
+	void run() override final {
+		throw std::runtime_error("run() is not supported by this algorithm; use run(trades)");
+	};
+
+	void run(const std::vector<std::pair<node, node> >& trades);
+
+	Graph getGraph(bool parallel = false);
+
+	virtual std::string toString() const override final;
+
+	virtual bool isParallel() const {
+		return false;
+	}
+
+	count getNumberOfAffectedEdges() const;
+
+
+private:
+	std::unique_ptr<CurveballDetails::CurveballIM> impl_;
```

Comment:
Why is a private implementation necessary here? There is only a single CurveballIM implementation, right? Why not just use that here without a heap allocation?

Reply:
The main reason to do so is to follow the pImpl-Idiom to provide a compilation barrier; this is useful since the .pyx file includes all public headers. If we do not use the barrier, a lot of Curveball specific header-files need to be parsed w/o benefit to the .pyx compilation process. The indirection won't affect the performance, since only a very small number of functions calls happen.